### PR TITLE
:sparkles: add support for Ethereum

### PR DESCRIPTION
**Pull request by menecio**
### Context
This PR adds support for the Ethereum network in the RatedAPI.

The proposal is that it should be used like this:
```python
import rated
from rated.networks import Ethereum

r = rated.Rated("ey...")
eth = r.ethereum(Ethereum.Mainnet)
for eff in eth.validator.effectiveness("0x123456789...", from_day=873, size=1): 
    print(f"Day: {eff.day}, Eff: {eff.validator_effectiveness}")

>>> Day: 873, Eff: 98.82005899705014
>>> Day: 872, Eff: 98.83015896502001
>>> ...
```

### CODE REVIEW

File: src/rated/ethereum/datatypes.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
+from dataclasses import dataclass
+from typing import Optional, List
+
+
+@dataclass
```
Comment (haiyanghe) [inline src/rated/ethereum/datatypes.py:5]: Are we using dataclasses instead of Pydantic models in this library?

**Review by haiyanghe — APPROVED** [review]

File: src/rated/ethereum/datatypes.py (hunk 1 above)
Comment (menecio) [reply src/rated/ethereum/datatypes.py:5]: We could use pydantic but it doesn't really add much other than having identical classes to the ones we already use. 

But if we think we should use pydantic, I don't have a strong opinion against it.

File: README.md [hunk 2]
```diff
@@ -6,3 +6,29 @@ Python bindings for the Rated API
...
+```
+
+## Usage
+**Example:** how to get a validator effectiveness rating by pubkey
+
+```python
+import rated
+from rated.networks import Ethereum
+
+r = rated.Rated("ey...")
```
Comment (ariskk) [inline README.md:24]: Some definition of what `ey...` is and how to get it (short description, point to console etc).

File: src/rated/__init__.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+from rated import ethereum
+from rated.client import Client
+from rated.networks import Ethereum
+
+api_key: Optional[str] = os.getenv("RATED_API_KEY")
+
+
+class Rated:
+    def __init__(self, api_key_: str | None = None):
+        self.api_key = api_key_ or api_key
```
Comment (ariskk) [inline src/rated/__init__.py:15]: if this is empty, there is no meaningful way to progress. So I guess better if we raise the `ValueError` here?

File: src/rated/ethereum/datatypes.py (hunk 1 above)
Comment (ariskk) [reply src/rated/ethereum/datatypes.py:5]: Slight preference for pydantic given the types are a bit more descriptive (eg `NonNegativeInt` vs just `int`)
What do you think @JavierHuertaGarces ?

File: src/rated/ethereum/datatypes.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    sum_late_source_penalties: Optional[float]
+    execution_proposed_empty_count: Optional[int]
+    sum_missed_attestation_rewards: Optional[float]
+    sum_missed_sync_committee_rewards: Optional[float]
+    sum_externally_sourced_execution_rewards: Optional[int]
+    day: Optional[int]
+    start_day: Optional[int]
+    end_day: Optional[int]
+    start_epoch: Optional[int]
+    end_epoch: Optional[int]
```
Comment (ariskk) [inline src/rated/ethereum/datatypes.py:85]: Something interesting we could do here is attach methods that thematically returns subsets. eg we could have a class for `ValidatorRewards` and then have the ability to do `effectiveness.rewards()`

File: src/rated/ethereum/validators.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    ValidatorAPR,
+    ValidatorMetadata,
+    ValidatorEffectiveness,
+)
+
+
+class Validator(APIResource):
+    path = "/validators"
+
+    def metadata(self, index_or_pubkey: int | str) -> ValidatorMetadata:
```
Comment (ariskk) [inline src/rated/ethereum/validators.py:17]: we should def have some validation here (can be pydantic or something else) to verify that  `index_or_pubkey` is either a `NonNegativeInt` or a 96/98 character long hex. That is to avoid sending to the API a request we know will fail.

File: src/rated/ethereum/validators.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        validator = self.client.get(f"{self.resource_path}/{index_or_pubkey}")
+        return json_to_instance(validator.json(), ValidatorMetadata)
+
+    def apr(
+        self,
+        index_or_pubkey: int | str,
+        *,
+        apr_type: str = "forward",
+        time_window: str = "1d",
```
Comment (ariskk) [inline src/rated/ethereum/validators.py:26]: Would be good to validate this before sending and raise a ValueError with the valid options

File: src/rated/client.py [hunk 7]
```diff
@@ -24,28 +23,33 @@ def raise_on_4xx_5xx(response: httpx.Response):
...
+            base_url=api_base_url,
             follow_redirects=True,
             event_hooks={"response": [raise_on_4xx_5xx]},
         )
 
+    @property
+    def headers(self) -> httpx.Headers:
+        return self._headers
+
     def get(self, *args, **kwargs) -> httpx.Response:
```
Comment (ariskk) [inline src/rated/client.py:49]: I wonder if a `stamina` config would be useful here

File: src/rated/ethereum/validators.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+        size: int | None = None,
+    ) -> Iterator[ValidatorEffectiveness]:
+        from_: str | int | date | None = from_day
+        if from_ is not None and isinstance(from_, date):
+            from_ = from_.isoformat()
+
+        params: Dict[str, Any] | None = {"from": from_, "size": size}
+        url = f"{self.resource_path}/{index_or_pubkey}/effectiveness"
+        while True:
+            effectiveness = self.client.get(url, params=params)
```
Comment (ariskk) [inline src/rated/ethereum/validators.py:49]: since we do `effectinveness.json()` three times, maybe assign it to a variable.

File: src/rated/ethereum/validators.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+        from_day: int | date | None = None,
+        size: int | None = None,
+    ) -> Iterator[ValidatorEffectiveness]:
+        from_: str | int | date | None = from_day
+        if from_ is not None and isinstance(from_, date):
+            from_ = from_.isoformat()
+
+        params: Dict[str, Any] | None = {"from": from_, "size": size}
+        url = f"{self.resource_path}/{index_or_pubkey}/effectiveness"
+        while True:
```
Comment (ariskk) [inline src/rated/ethereum/validators.py:48]: this will get all the data all the way to genesis, regardless of the `size` the user has submitted. Is that intended?

File: src/rated/__init__.py [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+
+    def ethereum(self, network: Ethereum) -> ethereum.Ethereum:
```
Comment (JavierHuertaGarces) [inline src/rated/__init__.py:17]: I'm not a fan of the network Enum and the actual network client sharing a name. There are a lot of `ethereum` and `Ethereum` around in packages, classes and types, which can make the code a bit hard to follow. Maybe we could be more specific in the naming.

File: src/rated/base.py [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+
+class Network:
+    path: str
+
+    def __init__(self, client: Client) -> None:
+        self._client = client
+
+    @property
+    def client(self) -> Client:
```
Comment (JavierHuertaGarces) [inline src/rated/base.py:15]: Not sure if this is a pattern we are purposely following, but why make the property within the `__init__` private if you are going to expose it here anyway?

File: src/rated/client.py [hunk 12]
```diff
@@ -24,28 +23,33 @@ def raise_on_4xx_5xx(response: httpx.Response):
 
 
 class Client:
-    def __init__(self, api_key: str | None = None, network: str = "mainnet"):
+    def __init__(self, api_key: str, network: NetworkEnum):
         self.api_key = os.getenv("RATED_API_KEY", api_key)
```
Comment (JavierHuertaGarces) [inline src/rated/client.py:27]: We should probably get rid of this here, as we already get the `RATED_API_KEY` before we call the client. Same thing with the `api_key` validation. I'd probably keep it in one place. If, for some reason, we want to keep it in both places, then I'd declare the env variable name somewhere to not repeat the string value.

File: src/rated/__init__.py (hunk 3 above)
Comment (menecio) [reply src/rated/__init__.py:15]: it will once you try to instantiate a network, which is when we init the `httpx.Client`

File: src/rated/ethereum/validators.py (hunk 8 above)
Comment (menecio) [reply src/rated/ethereum/validators.py:49]: yup, there is some cleaning needed, def going to do that

File: src/rated/ethereum/validators.py (hunk 9 above)
Comment (menecio) [reply src/rated/ethereum/validators.py:48]: yes, so that you iterate over all pages by default. The use consuming would be responsible to stop iterating, 
e.g.
```python
for eff in eth.validator.effectiveness("0x123456789...", from_day=873, size=1): 
    if eff.day < 800:
        break
    print(f"Day: {eff.day}, Eff: {eff.validator_effectiveness}")
```
wdyt?

File: src/rated/ethereum/datatypes.py (hunk 1 above)
Comment (JavierHuertaGarces) [reply src/rated/ethereum/datatypes.py:5]: My care factor for this is low, to be honest. I would say that unless we need to get that specific with types, I'd slightly favour dataclasses because you don't need an external dependency for them.

File: src/rated/__init__.py (hunk 10 above)
Comment (JavierHuertaGarces) [reply src/rated/__init__.py:17]: This might be a little out there, but what if you move the Ethereum network enum to the ethereum package? It's a bit more verbose, but everything comes from the same place, and packages from different networks would have their own `Network` enum. 

```python
import rated
from rated.ethereum import Network # It could also be aliased as EthNetwork for clarity

r = rated.Rated("ey...")

eth = r.ethereum(Network.Mainnet)
```

That also means that when creating the client:

```python
def ethereum(self, network: ethereum.Network) -> ethereum.Ethereum:
```

Comment (JavierHuertaGarces) [reply src/rated/__init__.py:17]: Also, we could change the enum values to uppercase so we don't confuse them with classes (something like https://docs.python.org/3/library/enum.html).

```python
import rated
from rated.ethereum import Network # It could also be aliased as EthNetwork for clarity

r = rated.Rated("ey...")

eth = r.ethereum(Network.MAINNET)
```

File: src/rated/ethereum/validators.py (hunk 9 above)
Comment (ariskk) [reply src/rated/ethereum/validators.py:48]: It seems counterintuitive. I wouldn't expect this to happen at all. I would expect to get one datapoint calling this `eth.validator.effectiveness("0x123456789...", from_day=873, size=1)`

Comment (menecio) [reply src/rated/ethereum/validators.py:48]: ok, I will add a `follow_next: bool = False` parameter, this will allow users to choose if they want to keep iterating if there is a next page

File: README.md [hunk 13]
```diff
@@ -6,3 +6,28 @@ Python bindings for the Rated API
...
+pip install rated-python
+```
+
+## Usage
+**Example:** how to get a validator effectiveness rating by pubkey
+
+```python
+from rated import Rated
+from rated.ethereum import MAINNET
+r = Rated("ey...")
```
Comment (ariskk) [inline README.md:23]: Still needs a short guide how to obtain the key. Also better if we explicitly say what `"ey..."` is, ie `API_KEY = 'ey...'`

File: src/rated/ethereum/datatypes.py [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+    sum_missed_attestations: Optional[int]
+    sum_missed_attestation_penalties: Optional[float]
+    sum_wrong_head_votes: Optional[int]
+    sum_wrong_head_penalties: Optional[float]
+    sum_attestation_rewards: Optional[float]
+    sum_late_source_penalties: Optional[float]
+    execution_proposed_empty_count: Optional[int]
+    sum_missed_attestation_rewards: Optional[float]
+    sum_missed_sync_committee_rewards: Optional[float]
+    sum_externally_sourced_execution_rewards: Optional[int]
```
Comment (ariskk) [inline src/rated/ethereum/datatypes.py:80]: Not all of them are optional, would it better if this mirrored the API's models in terms of optionality

File: src/rated/ethereum/slashings.py [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+            from_ = from_.isoformat()
+        params: Dict[str, Any] = {"from": from_, "size": size}
+        return self.client.yield_paginated_results(
+            url,
+            params=params,
+            cls=SlashingPenalty,
+            follow_next=follow_next,
+        )
+
+    def for_validator(
```
Comment (menecio) [inline src/rated/ethereum/slashings.py:73]: what do you think about this naming? @rated-network/devs 

```python
import rated


r = rated.Rated("some_key")
eth = r.ethereum(network=rated.ethereum.MAINNET)

# by validator index
slashings = eth.slashings.for_validator(500_000)

# by pubkey
slashings = eth.slashings.for_validator("0x...")

for penalty in slashings:
    print(penalty)
```

File: src/rated/__init__.py [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+
+
+class Rated:
+    def __init__(self, api_key: str | None = None):
+        self.api_key = api_key
+        if self.api_key is None:
+            self.api_key = os.getenv("RATED_API_KEY")
+
+    def ethereum(self, network: str) -> ethereum.Ethereum:
+        if self.api_key is None:
```
Comment (JavierHuertaGarces) [inline src/rated/__init__.py:16]: I would switch this if around for readability:
```
if self.api_key:
    c = Client(api_key=self.api_key, network=network)
    return ethereum.Ethereum(c)

raise ValueError("Must provide an API key.")
```

File: src/rated/base.py [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+
+from rated.client import Client
+
+
+class Network:
+    path: str
+    supported_networks: Sequence[str]
+
+    def __init__(self, client: Client) -> None:
+        if client.network not in self.supported_networks:
```
Comment (JavierHuertaGarces) [inline src/rated/base.py:11]: Would test for positive here as well.

File: src/rated/client.py [hunk 18]
```diff
@@ -24,28 +26,59 @@ def raise_on_4xx_5xx(response: httpx.Response):
 
 
 class Client:
-    def __init__(self, api_key: str | None = None, network: str = "mainnet"):
-        self.api_key = os.getenv("RATED_API_KEY", api_key)
+    def __init__(self, api_key: str, network: str):
+        self.api_key = api_key
+        self.network: str = network
         if not self.api_key:
```
Comment (JavierHuertaGarces) [inline src/rated/client.py:32]: Why do we recheck this? We already do in `class Rated:`

File: src/rated/base.py [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+
+    @property
+    def client(self) -> Client:
+        return self.network.client
+
+    @property
+    def resource_path(self) -> str:
+        return f"{self.network.path}{self.path}"
+
+    def get(self, *args, **kwargs) -> Any:
```
Comment (JavierHuertaGarces) [inline src/rated/base.py:30]: I find it a bit confusing having this here. I'd rather use the `get` in the client directly.

File: src/rated/ethereum/network.py [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+            yield json_to_instance(item, NetworkOverview)
+
+    def capacity(self) -> Iterator[NetworkChurnCapacity]:
+        data = self.get(f"{self.resource_path}/capacity")
+        for item in data:
+            yield json_to_instance(item, NetworkChurnCapacity)
+
+    def capacity_pool(
+        self,
+        stake_action: str = "activation",
```
Comment (JavierHuertaGarces) [inline src/rated/ethereum/network.py:33]: Could we have enums for all API parameters that have a restricted set of values?

File: src/rated/ethereum/validators.py (hunk 5 above)
Comment (JavierHuertaGarces) [reply src/rated/ethereum/validators.py:18]: Could we be more verbose and have 2 exclusive arguments? I think it will make validation easier to read, and the user can be more explicit in the function call.

**Review by JavierHuertaGarces — CHANGES_REQUESTED** [review]

File: src/rated/client.py [hunk 21]
```diff
@@ -24,28 +26,59 @@ def raise_on_4xx_5xx(response: httpx.Response):
...
+                if cls:
+                    yield json_to_instance(item, cls)
+                else:
+                    yield item
+
+            if not content["next"] or not follow_next:
+                break
+
+            url = content["next"]
+            params_ = None
```
[the remark this answers is not in the document]
Comment (menecio) [reply src/rated/client.py:79]: it comes embedded in the next url, so the params we carry from the previous requests are not useful anymore

File: src/rated/ethereum/enums.py [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+    NONE = None
+    DEPOSIT_ADDRESS = "depositAddress"
+    WITHDRAWAL_ADDRESS = "withdrawalAddress"
+    NODE_OPERATOR = "nodeOperator"
+    POOL = "pool"
+    POOL_SHARE = "poolShare"
+
+
+class Granularity(str, Enum):
+    NONE = None
```
Comment (JavierHuertaGarces) [inline src/rated/ethereum/enums.py:26]: Instead of having a `NONE` value for the enums that support it, couldn't we declare the arguments that use it as, in this case, `Granularity | None` instead?

**Review by JavierHuertaGarces — APPROVED** [review]

File: src/rated/ethereum/enums.py (hunk 22 above)
Comment (menecio) [reply src/rated/ethereum/enums.py:26]: it forces then to check `granularity.value if granularity is not None else None`, this way we can simply do granularity.value for any case
